components/service-operator/apis/storage/v1beta1/s3_types.go

```diff
@@ -268,3 +276,44 @@ func (s *S3Bucket) GetStackRoleParameters(roleName string) ([]*cloudformation.Pa
 	}
 	return params, nil
 }
+
+func (s *S3Bucket) Empty(client sdk.Client) error {
+	var err error
+	errOuter := client.ListObjectsV2Pages(
+		&s3.ListObjectsV2Input{
+			Bucket:  aws.String(s.GetAWSName()),
+		},
+		func (objects *s3.ListObjectsV2Output, lastPage bool) bool {
+			var objectIDs []*s3.ObjectIdentifier
+			for _, object := range objects.Contents {
+				objectIDs = append(objectIDs, &s3.ObjectIdentifier{Key: object.Key})
+			}
+
+			output, err := client.DeleteObjects(&s3.DeleteObjectsInput{
```

Comment:
can we move this outside of the "loop" as it's quite hard to read, and then avoids the weird `err` shadowing

Reply:
Possibly but DeleteObjects can only handle 1000 objects at a time and ListObjectsV2 only returns up to 1000 at a time, therefore right now we don't have to worry about batching calls to DeleteObjects. If we move the DeleteObjects call out and build a complete list of objects to delete and try to do them all in one go, we may exceed the limit.

Reply:
We spoke in person and whereas I envisaged a single flat list of all objects to delete (which we'd then need to split into batches), Chris has pointed out I can just make a list of lists. Dunno why I didn't think of that, doing.